### ISSUE TITLE
Allow emap to accept data with pathway ID

### ISSUE THREAD

**Issue by jvwong** [PathwayCommons/app-ui#563]
## Description of new feature

The emap service accepts a list of pathwayIDs. It uses this information to calculates pair-wise similarity between pathways and outputs as a graph. It would be valuable to allow input of pathway (i.e. node)-related data that can be passed into the output, and used by the client. 

## Motivation for new feature

One important example is `pvalue` etc from the enrichment service output.

```js
{
	"GO:0006354": {
		...
		"pvalue": "2.65e-02",
		"tType": "BP",
		"tName": "   DNA-templated transcription, elongation",
		"Q&TList": "PAF1,ZMYND11,POLR2B,AFF4,SUPT16H,ELP4,GTF2H1,BRD4,SSRP1,SETD2"
	},
...
}
```

**Comment by wxli0** [comment]
on branch emap-data
input:
pathwayIdList=GO:0043525 GO:0043523 pvalue 2.65e-02 GO:0072395 tType BP pvalue 2.64e-02 GO:0072401&cutoff=0.3&OCWeight=0.8
- delimited by whitespace
- strings not pathway IDs are related data for the previous pathway ID
- throw error if input doesn't follow the specified format
output:
```
{
	"unrecognized": [],
	"duplicate": [],
	"cutoff": "0.3",
	"JCWeight": "0.2",
	"OCWeight": 0.8,
	"graph": {
		"nodes": {
			"GO:0043525": {}
			"GO:0043523": {"pvalue": 2.65e-02}
                         "GO:0072395": {"tType": BP, "pvalue": 2.64e-02 }
                          "GO:0072401":{}
		},
		"edges": [
			{
				"id": "GO:0043525_GO:0043523",
				"source": "GO:0043525",
				"target": "GO:0043523",
				"similarity": 0.8517766497461929,
				"intersection": [
					...
				]
			}
		]
	}
}
```

**Comment by wxli0** [comment]
Since we allow duplicate in emap service, how to handle the following cases?
1. pathwayIdList = GO:0043523 pvalue 1 GO:0043523 pvalue 2
- report error? 
2. pathwayIdList = GO:0043523 pvalue 2 GO:0043523 tType BP
- combine two properties (pvalue, tType) with GO:0043523?

**Comment by jvwong** [comment]
I think it's best to report error in both cases, that is, when you have non-unique IDs.

**Comment by jvwong** [comment]
Sort of related, does it make sense to have the "graph" object look like JSON that could be directly inserted to cytoscape.js, inside ["elements"](http://js.cytoscape.org/#notation/elements-json)? 

That is, given your response above, would the client need another function to map "graph" into something suitable for cytoscape.js?  

http://js.cytoscape.org/#notation/elements-json

**Comment by wxli0** [comment]
![screen shot 2018-03-27 at 9 18 02 am](https://user-images.githubusercontent.com/24982880/37969828-cdce798e-319f-11e8-86fe-ff183665253c.png)
The current logic for emap service is that it detects duplicates and unrecognized, removes them from pathwayIdList, reports them and continues processing. Do you mean we should throw an error and terminate instead?

**Comment by wxli0** [comment]
> Sort of related, does it make sense to have the "graph" object look like JSON that could be directly inserted to cytoscape.js, inside "elements"?
> 
> That is, given your response above, would the client need another function to map "graph" into something suitable for cytoscape.js?

will make an issue for that.

**Comment by wxli0** [comment]
I think this also relates to what to do with unrecognized and duplicate pathway IDs without additional data. Here are some possible ways:
1. (currently using)
unrecognized -> remove, push onto unrecognized array, continue
duplicate -> remove, push onto duplicate array, continue
2. 
unrecognized -> throw error, no unrecognized array in output
duplicate -> remove, push onto duplicate array, continue
3. 
unrecognized -> throw error
duplicate -> throw error
no duplicate and unrecognized array in output

3. makes it easier when it comes to importing additional data while 1. seems more compatible with input list.

**Comment by wxli0** [comment]
Is this one valid? Can GO/REACTOME be the value for an additional attribute?
`pathwayIdList = GO:0043523 attr GO:0043522 GO:0043523`

We can say GO:0043522 is the value for attr, so it's valid.
Or we can say there is no value for attr, hence invalid.

Or is it better to come up with another format to specify additional data?
e.g. `pathwayIdList = GO:0043523 attr:val GO:0043522 pvalue:2 GO:0043523`

**Comment by wxli0** [comment]
input format:
pathwayIdList:
```
[
"GO:0006354": {
			"pValue": 0.0087,
			"t": 124,
			"q": 1,
			"qAndT": 1,
			"qAndTOverQ": 1,
		},
"GO:0006368": {
			"pValue": 0.0087,
			"t": 100,
			"tName": "transcription elongation from RNA polymerase II promoter",
			"tDepth": 2,
			"qAndTList": [
				"AFF4"
			]
		},
"REAC:674695": {}
]
```

**Comment by wxli0** [comment]
wait on pr #586 #580

**Comment by jvwong** [comment]
I would say yes, as we discussed, can let the client push in additional fields they want passed to the cytoscape-JSON output.

**Comment by jvwong** [comment]
Rather than create a separate issue I will leave this here: I find the input to `emap` very awkward. I think a lot of the problems arise from the fact that your `enrichment` output looks very different than your `emap` input format. Wouldn't it be easier for `emap` to just accept JSON object that looks more or less like the output from `enrichment`
```js
{
	"options": {
		"cutoff": 0.375,
		"JCWeight": 0.5,
		"OCWeight": 0.5
	},
	"pathwayInfo": {
		"GO:0006354": {
			"pValue": 0.0274,
			"t": 18,
			"q": 58,
			"qAndT": 10,
			"qAndTOverQ": 0.172,
			"qAndTOverT": 0.556,
			"tType": "BP",
			"tGroup": 1,
			"tName": "DNA-templated transcription, elongation",
			"tDepth": 1,
			"qAndTList": [
				"PAF1",
				"ZMYND11",
				"POLR2B",
				"AFF4",
				"SUPT16H",
				"ELP4",
				"GTF2H1",
				"BRD4",
				"SSRP1",
				"SETD2"
			]
		},
		"GO:0006368": {
			"pValue": 0.0274,
			"t": 18,
			"q": 58,
			"qAndT": 10,
			"qAndTOverQ": 0.172,
			"qAndTOverT": 0.556,
			"tType": "BP",
			"tGroup": 1,
			"tName": "transcription elongation from RNA polymerase II promoter",
			"tDepth": 1,
			"qAndTList": [
				"PAF1",
				"ZMYND11",
				"POLR2B",
				"AFF4",
				"SUPT16H",
				"ELP4",
				"GTF2H1",
				"BRD4",
				"SSRP1",
				"SETD2"
			]
		}
	}
}
```

**Comment by wxli0** [comment]
If we do it this way, then we can't detect the duplicate case of pathwayIds #585 since there can not be duplicate keys in an object, latter values will always override previous ones. 
e.g. {a: 'a', a: 'n'} => {a: 'n'}

I'm not sure if it's acceptable to hide duplicate errors if users want to create the input on their own.

**Comment by jvwong** [comment]
I would say that hiding the duplicate error is an acceptable tradeoff to allow our enrichment output to be used as input to emap, given that a JS object should not/cannot have duplicate keys.

**Comment by wxli0** [comment]
Ok. will make a note in the documentation.

**Comment by wxli0** [comment]
haven't committed rest.js since it's dependent on pr #611
